### ISSUE TITLE
Installer fails on cygwin with git and chsh errors

### ISSUE THREAD

**Issue by apjanke** [robbyrussell/oh-my-zsh#3588]
I tried doing a fresh install of oh-my-zsh in Cygwin on Windows 7 x64, and it failed to install, with kind of cryptic messages. Both my Windows and cygwin installations are up to date as of 2/14/2015.

If you have Git for Windows installed and on the path, but not Cygwin's git installed, `install.sh` will use the Windows git, and for me it failed to clone the OMZ repo. Looks like Cygwin's git must be installed for `install.sh` to work.

If `install.sh` makes it past the repo cloning, it will then fail when trying to change the shell, because `chsh` does not exist on Cygwin.

##  What the failure looks like

Here's me running install when Windows Git 1.9.4 but not Cygwin git is installed.

```
gwydion% which git
/cygdrive/c/Program Files (x86)/Git/cmd/git
gwydion% git --version
git version 1.9.4.msysgit.0
gwydion% curl -L http://install.ohmyz.sh | sh
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   115  100   115    0     0    149      0 --:--:-- --:--:-- --:--:--   393
100  1930  100  1930    0     0   2014      0 --:--:-- --:--:-- --:--:--  2014
\033[0;34mCloning Oh My Zsh...\033[0m
Cloning into '/home/janke/.oh-my-zsh'...
\033[0;34mLooking for an existing zsh config...\033[0m
\033[0;33mFound ~/.zshrc.\033[0m \033[0;32mBacking up to ~/.zshrc.pre-oh-my-zsh\033[0m
\033[0;34mUsing the Oh My Zsh template file and adding it to ~/.zshrc\033[0m
cp: cannot stat ‘/home/janke/.oh-my-zsh/templates/zshrc.zsh-template’: No such file or directory
gwydion%
```

With Windows Git 1.9.5 installed, it fails at the same point but you get a different error message.

```
gwydion% curl -L http://install.ohmyz.sh | sh
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   115  100   115    0     0     80      0  0:00:01  0:00:01 --:--:--   155
100  1930  100  1930    0     0    643      0  0:00:03  0:00:03 --:--:--  1783
\033[0;34mCloning Oh My Zsh...\033[0m
fatal: destination path '/home/janke/.oh-my-zsh' already exists and is not an empty directory.
git not installed
gwydion%
```

Despite what the error message says, '/home/janke/.oh-my-zsh' did not in fact exist.

I think what's going on here is that git 1.9.5 is now returning an exit status in this case. The code in install.zsh is lumping any error from the git invocation together and reporting it as not being installed. This should probably be two separate checks.
```
hash git >/dev/null 2>&1 && env git clone --depth=1 https://github.com/robbyrussell/oh-my-zsh.git $ZSH || {
  echo "git not installed"
  exit
}
```

I was able to get the clone to work using Windows git if I gave it a relative path to clone to. May be something with how Windows git is handling Unix style paths.
```
gwydion% env git clone --depth=1 https://github.com/robbyrussell/oh-my-zsh.git $ZSH
fatal: destination path '/home/janke/.oh-my-zsh' already exists and is not an empty directory.
gwydion% file /home/janke/.oh-my-zsh
/home/janke/.oh-my-zsh: cannot open `/home/janke/.oh-my-zsh' (No such file or directory)
gwydion% env git clone --depth=1 https://github.com/robbyrussell/oh-my-zsh.git .oh-my-zsh
Cloning into '.oh-my-zsh'...
gwydion%
```

Installing Cygwin git fixes that issue. After that, it will get most of the way through the install, and then break when trying to call chsh.
```
gwydion% which git
/usr/bin/git
gwydion% curl -L http://install.ohmyz.sh | sh
...
\033[0;34mCloning Oh My Zsh...\033[0m
Cloning into '/home/janke/.oh-my-zsh'...
...
\033[0;34mLooking for an existing zsh config...\033[0m
\033[0;33mFound ~/.zshrc.\033[0m \033[0;32mBacking up to ~/.zshrc.pre-oh-my-zsh\033[0m
\033[0;34mUsing the Oh My Zsh template file and adding it to ~/.zshrc\033[0m
\033[0;34mCopying your current PATH and adding it to the end of ~/.zshrc for you.\033[0m
\033[0;34mTime to change your default shell to zsh!\033[0m
sh: line 37: chsh: command not found
gwydion%

## What to do?

Since running the installer on cygwin has some prerequisites, we should probably document them in the installation instructions. And maybe we could change `install.zsh` to handle the cygwin cases by:
* Detecting if the `git` it found is the Windows/MSYS Git and throwing an error message asking for Cygwin git to be installed
* Testing whether `chsh` exists on the system, and giving a "please change your shell manually" message if it doesn't

Also, it seems the installer should exit with a nonzero status when it's aborting due to an error, like git missing or OMZ already being installed.

**Comment by apjanke** [comment]
After poking around my drive, I found a C:\home\janke\.oh-my-zsh directory with the repo cloned in to it. Looks like Windows git is treating the paths as Windows paths and not Unix paths (under cygwin's filesystem view). Which of course makes sense.

We could detect Cygwin+ Windows Git and use `cygpath` to convert the paths, but that sounds like a big pain, and it would have to be done everywhere. Should probably just require cygwin git, or just a git that works natively within the environment that oh-my-zsh is calling it from.

**Comment by apjanke** [comment]
Got a fix for this going on branch https://github.com/apjanke/oh-my-zsh/tree/installer-cygwin-support. Will test it on a couple other platforms before making a PR out of it.

**Comment by mcornella** [comment]
I assume #3550 is related to this one?

**Comment by mcornella** [comment]
Nevermind, it's #3597 the one that fixes #3550. Didn't see it. I'll take a closer look at this one later tomorrow

**Comment by apjanke** [comment]
Not related to #3550; I just discovered it when I was setting up OMZ on Cygwin to work on #3550.

I think the git issue is just that Windows/MSYS Git is a Windows program that takes normal Windows paths instead of Unix-style Cygwin paths. The path `/home/janke/.oh-my-zsh`, when interpreted as a Windows path, is relative to the current drive letter, and is unaware of Cygwin's Unix-style filesystem root. So the initial `git` command clones the repo to `C:\home\janke\.oh-my-zsh`, succeeds, and returns 0. When I upgraded Windows/MSYS Git to 1.9.5 and ran it again, it saw the `C:\home\janke\.oh-my-zsh` from the last run and errored out. The idea that 1.9.5 was returning different error statuses was a red herring.
